compiler_opt/rl/corpus.py

```diff
@@ -104,194 +259,161 @@ def __init__(self,
+      else:
+        if fthinlto_index_flag in replace_flags:
+          raise ValueError(
+              '-fthinlto-index must be handled by the infrastructure')
+        replace_flags[fthinlto_index_flag] = '{context.thinlto_full_path}'
+
+    additional_flags = ('-x', 'ir',
+                        '{context.module_full_path}') + additional_flags
+
+    # don't use add/remove for replace
+    add_keys = set(k.split('=', maxsplit=1)[0] for k in additional_flags)
+    if add_keys.intersection(
+        set(replace_flags)) or set(delete_flags).intersection(
+            set(replace_flags)) or add_keys.intersection(set(delete_flags)):
+      raise ValueError('do not use add/delete flags to replace')
+
+    self._additional_flags = additional_flags
+    self._delete_flags = delete_flags
+    self._replace_flags = replace_flags
+
+  def sample(self, k: int, sort: bool = False) -> List[ModuleSpec]:
+    """Samples `k` module_specs, optionally sorting by size descending.
+
+    Use load_corpus_element to get LoadedModuleSpecs - this allows the user
+    decide how the loading should happen (e.g. may want to use a threadpool)
     """
-    self._module_specs = tuple(
-        sorted(
-            _build_modulespecs_from_datapath(
-                data_path=data_path,
-                additional_flags=additional_flags,
-                delete_flags=delete_flags,
-                replace_flags=replace_flags),
-            key=lambda m: m.size,
-            reverse=True))
-    self._root_dir = data_path
-
-  @classmethod
-  def from_module_specs(cls, module_specs: Iterable[ModuleSpec]):
-    """Construct a Corpus from module specs. Mostly for testing purposes."""
-    cps = cls.__new__(cls)  # Avoid calling __init__
-    super(cls, cps).__init__()
-    cps._module_specs = tuple(
-        sorted(module_specs, key=lambda m: m.size, reverse=True))
-    cps.root_dir = None
-    return cps
-
-  def sample(self,
-             k: int,
-             sort: bool = False,
-             sampler=SamplerBucketRoundRobin()) -> List[ModuleSpec]:
-    """Samples `k` module_specs, optionally sorting by size descending."""
     # Note: sampler is intentionally defaulted to a mutable object, as the
     # only mutable attribute of SamplerBucketRoundRobin is its range cache.
-    k = min(len(self._module_specs), k)
+    k = min(len(self._contents), k)
     if k < 1:
       raise ValueError('Attempting to sample <1 module specs from corpus.')
-    sampled_specs = sampler(self._module_specs, k=k)
+    sampled_specs = self._sampler(self._contents, k=k)
     if sort:
       sampled_specs.sort(key=lambda m: m.size, reverse=True)
     return sampled_specs
 
-  def filter(self, p: re.Pattern):
-    """Filters module specs, keeping those which match the provided pattern."""
-    return Corpus.from_module_specs(
-        (ms for ms in self._module_specs if p.match(ms.name)))
+  def load_module_spec(self, module_spec: ModuleSpec) -> LoadedModuleSpec:
```

Comment:
Looks like should be the init of LoadModuleSpec class?

Reply:
The idea is to keel LoadedModuleSpec just a plain data object, decoupled from provenance. Later, if we want to (for example) fetch corpus from a database, we just need to change how we make LoadedModuleSpecs.

Reply:
i don't see a lot of benefit in keeping LoadModuleSpec a plain data object, and it already has some functions in it, making it as init will help make the API cleaner. But I'm fine with the current setup :)